File: snap_drawing/src/snap_drawing/cpp/Touches/PinchGestureRecognizer.cpp
//
//  PinchGestureRecognizer.cpp
//  valdi-skia
//
//  Created by Simon Corsin on 6/29/20.
//
#include "snap_drawing/cpp/Touches/PinchGestureRecognizer.hpp"
#include "snap_drawing/cpp/Touches/DragGestureRecognizer.hpp"
#include "snap_drawing/cpp/Touches/RotateGestureRecognizer.hpp"

#include "utils/debugging/Assert.hpp"
#include <cmath>

namespace snap::drawing {

PinchGestureRecognizer::PinchGestureRecognizer(const GesturesConfiguration& /*gesturesConfiguration*/) {}

PinchGestureRecognizer::~PinchGestureRecognizer() = default;

bool PinchGestureRecognizer::requiresFailureOf(const GestureRecognizer& other) const {
    return dynamic_cast<const PinchGestureRecognizer*>(&other) != nullptr;
}

bool PinchGestureRecognizer::canRecognizeSimultaneously(const GestureRecognizer& other) const {
    if (dynamic_cast<const RotateGestureRecognizer*>(&other) != nullptr) {
        return true;
    }
    if (dynamic_cast<const DragGestureRecognizer*>(&other) != nullptr) {
        return true;
    }
    return false;
}

PinchEvent PinchGestureRecognizer::makeMoveEvent() const {
    PinchEvent pinchEvent = MoveGestureRecognizer<PinchEvent>::makeMoveEvent();
    if (pinchEvent.pointerCount > 1 && _moveState.value().currentEvent.getType() != TouchEventTypePointerUp &&
        _moveState.value().currentEvent.getType() != TouchEventTypePointerDown) {
        // for events with more than one pointer, we use the delta from the cached value
        pinchEvent.scale = getCurrentScale() * _netScale;
    } else {
        // currently, our gesture detection system fires one last event for the finger leaving with a null currentEvent,
        // and we want to ensure any changes are persisted
        pinchEvent.scale = _netScale;
    }

    return pinchEvent;
}

// (mli6) WARN) - startEvent setting for multitouch does not track perfectly accurately for resumptions -
// Ticket: 2885
Scalar PinchGestureRecognizer::getCurrentScale() const {
    auto startDirection = _moveState.value().startEvent.getDirection();
    auto currentDirection = _moveState.value().currentEvent.getDirection();

    auto startScale = Point::length(startDirection.dx, startDirection.dy);
    auto currentScale = Point::length(currentDirection.dx, currentDirection.dy);

    return currentScale / startScale;
}

bool PinchGestureRecognizer::shouldStartMove(const TouchEvent& event) const {
    return event.getPointerCount() > 1;
}

bool PinchGestureRecognizer::shouldContinueMove(const TouchEvent& event) const {
    return event.getPointerCount() > 0;
}

void PinchGestureRecognizer::onPointerChange(const TouchEvent& event) {
    if (event.getPointerCount() == 2 && event.getType() == TouchEventTypePointerUp) {
        // if any touches are still present, we allow for a scale to continue
        // when end users lift all but one finger, we cache the current scale amount such that they can resume the
        // scale

        // we only cache this when the pointer count decreases, such that multifinger events or duplicate events don't
        // cause a re-caching
        _netScale *= getCurrentScale();
        transitionToState(GestureRecognizerStateChanged);
    }
}

void PinchGestureRecognizer::onReset() {
    MoveGestureRecognizer::onReset();
    _netScale = 1;
}

std::string_view PinchGestureRecognizer::getTypeName() const {
    return "pinch";
}
} // namespace snap::drawing

File: valdi/src/java/com/snap/valdi/views/touches/PinchGestureRecognizer.kt
package com.snap.valdi.views.touches

import android.view.MotionEvent
import android.view.View
import android.view.ScaleGestureDetector

interface PinchGestureRecognizerListener {
    fun onRecognized(gesture: ValdiGestureRecognizer, state: ValdiGestureRecognizerState, x: Int, y: Int, scale: Float, pointerCount: Int, pointerLocations: List<ValdiGesturePointer>)
    fun shouldBegin(gesture: ValdiGestureRecognizer, x: Int, y: Int, scale: Float, pointerCount: Int, pointerLocations: List<ValdiGesturePointer>): Boolean
}

/**
 * Deprecated - use PinchGestureRecognizerV2 instead
 */
open class PinchGestureRecognizer(
        view: View,
        val listener: PinchGestureRecognizerListener
) : ValdiGestureRecognizer(view) {

    private var currScale = 1f

    // track previous scales - such that multiple single scales can be all multiplied together
    private var netScale = 1f

    // we split this from pointerCount in ValdiGestureRecognizer to store the curr pointer count of the current event
    private var currPointerCount = 0

    // WARN (mli6) - this may be a cause for gesture start delays for pinch - Ticket: 2854
    private val gestureDetector = ScaleGestureDetector(view.context, object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
        override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
            if (state == ValdiGestureRecognizerState.POSSIBLE) {
                updateState(ValdiGestureRecognizerState.BEGAN)
            }
            return super.onScaleBegin(detector)
        }

        override fun onScale(detector: ScaleGestureDetector): Boolean {
            currScale = detector.scaleFactor
            return super.onScale(detector)
        }

        override fun onScaleEnd(detector: ScaleGestureDetector) {
            super.onScaleEnd(detector)
            // if any touches are still present, we allow for a scale to continue
            // when end users lift all but one finger, we cache the current scale amount such that they can resume the scale
            if (currPointerCount > 0) {
                netScale *= currScale
                currScale = 1f
            }

            val pointersRequired = 2
            updateState(
                if (pointerCount >= pointersRequired) ValdiGestureRecognizerState.CHANGED
                else ValdiGestureRecognizerState.ENDED
            )
        }
    })

    init {
        gestureDetector.isQuickScaleEnabled = false
    }

    override fun onUpdate(event: MotionEvent) {
        currPointerCount = event.pointerCount

        gestureDetector.onTouchEvent(event)
    }

    override fun onProcess() {
        listener.onRecognized(this, state, x, y, currScale * netScale, pointerCount, pointerLocations)
    }

    override fun shouldBegin(): Boolean {
        return listener.shouldBegin(this, x, y, currScale * netScale, pointerCount, pointerLocations)
    }

    override fun onReset(event: MotionEvent) {
        //Need to ensure ScaleDetector receives the up event so it can reset its state
        gestureDetector.onTouchEvent(event)

        super.onReset(event)
        currPointerCount = 0
        netScale = 1f
        currScale = 1f
    }

    override fun canRecognizeSimultaneously(other: ValdiGestureRecognizer): Boolean {
        return other.javaClass == DragGestureRecognizer::class.java ||
                other.javaClass == RotateGestureRecognizer::class.java ||
                other.javaClass == RotateGestureRecognizerV2::class.java
    }
}
